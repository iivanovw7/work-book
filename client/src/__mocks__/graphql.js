import {
  GET_POST, GET_POSTS, GET_TAGS, GET_USER, FIND_POSTS_BY_TAG, FIND_POSTS_BY_KEYWORD
} from '../graphql/queries';
import { DELETE_POST, ADD_POST, UPDATE_POST } from '../graphql/mutations';

export const gqlMocks = [
  /**
	 * [0] Returned after success GET POST request
	 */
  {
    request: {
      query: GET_POST,
      variables: {
        _id: '1'
      }
    },
    result: {
      data: {
        getPost: {
          _id: '1',
          text: 'Post text',
          title: 'Post title',
          subject: 'Post subject',
          created: 'Post date',
          author: 'Post author',
          tags: ['1', '2', '3']
        }
      }
    }
  },
  /**
	 * [1] Returned after errored GET POST request
	 */
  {
    request: {
      query: GET_POST,
      variables: {
        _id: '1'
      }
    },
    error: new Error('aw shucks')
  },
  /**
	 * [2] Returned after success DELETE POST mutation
	 */
  {
    request: {
      query: DELETE_POST,
      variables: {
        _id: '1'
      }
    },
    result: {
      deletePost: {
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    }
  },
  /**
	 * [3] Returned after errored DELETE POST mutation
	 */
  {
    request: {
      query: DELETE_POST,
      variables: {
        _id: '1'
      }
    },
    error: new Error('aw shucks')
  },
  /**
	 * [4] Returned after success ADD POST mutation
	 */
  {
    request: {
      query: ADD_POST,
      variables: {
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    },
    result: {
      addPost: {
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    }
  },
  /**
	 * [5] Returned after errored ADD POST mutation
	 */
  {
    request: {
      query: ADD_POST,
      variables: {
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    },
    error: new Error('aw shucks')
  },
  /**
	 * [6] Returned after success UPDATE POST mutation
	 */
  {
    request: {
      query: UPDATE_POST,
      variables: {
        _id: '1',
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    },
    result: {
      data: {
        getPost: {
          _id: '1',
          text: 'Post text',
          title: 'Post title',
          subject: 'Post subject',
          created: 'Post date',
          author: 'Post author',
          tags: ['1', '2', '3']
        }
      }
    }
  },
  /**
	 * [7] Returned after success UPDATE POST mutation
	 */
  {
    request: {
      query: UPDATE_POST,
      variables: {
        _id: '1',
        text: 'Post text',
        title: 'Post title',
        subject: 'Post subject',
        tags: ['1', '2', '3']
      }
    },
    error: new Error('aw shucks')
  },
  /**
	 * [8] Returned after success GET POSTS query
	 */
  {
    request: {
      query: GET_POSTS
    },
    result: {
      data: {
        getPosts: [
          {
            _id: '5d283694ddc2cb244f84bdca',
            author: '5cbeba32fb6fc0265f273c94',
            created: '1562915551933',
            subject: 'Cross browser solution for handling mouse wheel events in Angular with Rx.js for',
            tags: ['1', '2', '3'],
            text: 'Post text',
            title: 'Custom Angular mouseWheel scroll directive'
          }
        ]
      }
    }
  },
  /**
	 * [9] Returned after error after GET POSTS query
	 */
  {
    request: {
      query: GET_POSTS
    },
    error: new Error('aw shucks')
  },
  /**
	 * [10] Returned after success GET TAGS query
	 */
  {
    request: {
      query: GET_TAGS
    },
    result: {
      data: {
        getTags: ['1', '2', '3']
      }
    }
  },
  /**
	 * [11] Returned after errored GET TAGS query
	 */
  {
    request: {
      query: GET_TAGS
    },
    error: new Error('aw shucks')
  },
  /**
	 *  [12] Returned after success GET USER query
	 */
  {
    request: {
      query: GET_USER,
      variables: {
        _id: '1'
      }
    },
    result: {
      data: {
        getUser: {
          _id: '1',
          email: 'Email',
          phone: 'Phone',
          name: 'Name',
          surname: 'Surname',
          created: 'Date',
          role: 'Role'
        }
      }
    }
  },
  /**
	 *  [13] Returned after errored GET USER query
	 */
  {
    request: {
      query: GET_USER,
      variables: {
        _id: '1'
      }
    },
    error: new Error('aw shucks')
  },
  /**
   *  [14] Returned after FIND_POSTS_BY_TAG query
   */
  {
    request: {
      query: FIND_POSTS_BY_TAG,
      variables: {
        tag: '1'
      }
    },
    result: {
      data: {
        findPostsByTag: [
          {
            _id: '5d283694ddc2cb244f84bdca',
            author: '5cbeba32fb6fc0265f273c94',
            created: '1562915551933',
            subject: 'Cross browser solution for handling mouse wheel events in Angular with Rx.js for',
            tags: ['1', '2', '3'],
            text: 'Post text',
            title: 'Custom Angular mouseWheel scroll directive'
          }
        ]
      }
    }
  },
  /**
   * [15] Returned after errored FIND_POSTS_BY_TAG query
   */
  {
    request: {
      query: FIND_POSTS_BY_TAG,
      variables: {
        tag: '1'
      }
    },
    error: new Error('aw shucks')
  },
  /**
   *  [16] Returned after FIND_POSTS_BY_KEYWORD query
   */
  {
    request: {
      query: FIND_POSTS_BY_KEYWORD,
      variables: {
        keyword: 'Custom'
      }
    },
    result: {
      data: {
        findPostsByKeyword: [
          {
            _id: '5d283694ddc2cb244f84bdca',
            author: '5cbeba32fb6fc0265f273c94',
            created: '1562915551933',
            subject: 'Cross browser solution for handling mouse wheel events in Angular with Rx.js for',
            tags: ['1', '2', '3'],
            text: 'Post text',
            title: 'Custom Angular mouseWheel scroll directive'
          }
        ]
      }
    }
  },
  /**
   * [17] Returned after errored FIND_POSTS_BY_KEYWORD query
   */
  {
    request: {
      query: FIND_POSTS_BY_KEYWORD,
      variables: {
        keyword: 'Custom'
      }
    },
    error: new Error('aw shucks')
  }
];
