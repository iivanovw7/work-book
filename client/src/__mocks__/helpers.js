import * as constants from '../constants';

export const convertTimestampOptions = [
  [1305593400, constants.TIMESTAMP_UNITS.MS, '{YYYY} MM-DD SSS [Z] A'],
  [1305593400, constants.TIMESTAMP_UNITS.SEC, '{YYYY} MM-DD SSS [Z] A'],
  [undefined, constants.TIMESTAMP_UNITS.MS, 'MMMM DD, YYYY'],
  [null, 'units', null]
];

export const convertTimestampExpected = [
  '{1970} 01-16 000 Z AM',
  '{2011} 05-17 000 Z AM',
  'Invalid Date',
  'Invalid Date'
];
