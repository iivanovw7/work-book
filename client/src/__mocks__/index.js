export * from './loginResponse';
export * from './localizedText';
export * from './graphql';
