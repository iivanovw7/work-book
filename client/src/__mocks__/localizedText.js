import * as localized from '../assets/locales.json';

// Returns localised text for UI elements
export const localizedText = {
  locales: localized.locales,
  login: localized.login,
  navigation: localized.navigation,
  pageNotFound: localized.pageNotFound
};
