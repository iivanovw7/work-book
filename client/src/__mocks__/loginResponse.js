export const loginResponse = {
  // eslint-disable-next-line max-len
  jwt: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZDA1ZmY0YTMyM2ZjNDZjMmE4YWMxNzUiLCJpZCI6IjVkMDVmZjRhMzIzZmM0NmMyYThhYzE3NSIsImlhdCI6MTU2MTAzMTk4MiwiZXhwIjoxNTYzNjIzOTgyfQ.ldBlLabhjPJg3-TwvuE5MwfMnhLRk7lbBm22KNfV4ws',
  user: {
    _id: '5d05ff4a323fc46c2a8ac175',
    confirmed: true,
    blocked: false,
    username: 'manager',
    email: '9iiiuk3@gmail.com',
    provider: 'local',
    createdAt: '2019-06-16T08:35:22.613Z',
    updatedAt: '2019-06-16T08:35:22.771Z',
    __v: 0,
    id: '5d05ff4a323fc46c2a8ac175',
    role: {
      _id: '5d052d6f0f0d3e3186cd34a4',
      name: 'Authenticated',
      description: 'Default role given to authenticated user.',
      type: 'authenticated',
      __v: 0
    }
  }
};
