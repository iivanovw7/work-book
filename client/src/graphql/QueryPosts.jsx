import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { Query } from 'react-apollo';
import { withRouter } from 'react-router-dom';
import { GET_POSTS } from './queries';
import ErrorMessage from '../components/errorMessage';
import PostsList from '../components/Posts/List';
import Spinner from '../components/UI/Spinner';

const QueryPosts = (props) => {
  const { skip, limit } = props;
  const [skipped, setSkipped] = useState(skip);

  return (
    <Query query={GET_POSTS} variables={{ skip, limit }} fetchPolicy="cache-and-network">
      {({
        loading, error, data, fetchMore
      }) => {
        /**
         * Triggers fetching additional results and updating current query results
         * @param {Number} skipRecords - results to be skipped
         * @param {Number} queryLimit - number of results to be fetched
         * @return {*}
         */
        function triggerFetchMore(skipRecords, queryLimit) {
          return fetchMore({
            variables: { skip: skipRecords, limit: queryLimit },
            updateQuery: (previousResult = {}, { fetchMoreResult }) => {
              if (!fetchMoreResult || (skipped === skipRecords)) return previousResult;

              const previousEntries = previousResult.getPosts.posts;
              const newEntries = fetchMoreResult.getPosts.posts;
              setSkipped(skipRecords);

              return {
                ...previousResult,
                getPosts: {
                  ...fetchMoreResult.getPosts,
                  posts: [...previousEntries, ...newEntries]
                }
              };
            }
          });
        }

        if (data && data.getPosts) {
          const postsLength = data.getPosts.posts.length;
          const hasMore = (postsLength < data.getPosts.count);

          return (
            <PostsList
              data={data}
              hasMore={hasMore}
              loading={loading}
              loader={<Spinner theme={props.theme} margin="1em 0 0 0" />}
              atBottomHandler={() => {
                if (!loading && hasMore) {
                  triggerFetchMore(postsLength, limit);
                }
              }}
              {...props}
            />
          );
        }

        if (loading) return <Spinner theme={props.theme} margin="5em 0 0 0" />;
        if (error) {
          return (
            <ErrorMessage
              theme={props.theme}
              text="Back"
              message="Error =( Try again later..."
              handleClick={() => {
                props.history.push('/');
              }}
            />
          );
        }
      }}
    </Query>
  );
};

QueryPosts.defaultProps = {
  skip: 0,
  limit: 10
};

QueryPosts.propTypes = {
  history: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired,
  skip: PropTypes.number,
  limit: PropTypes.number
};

export default withRouter(QueryPosts);
