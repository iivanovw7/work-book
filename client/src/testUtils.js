import { createMemoryHistory } from 'history';
import * as localized from './assets/locales.json';

// Mock browser history
export const history = createMemoryHistory('/dashboard');

// Grid config object for testing
export const mappedSizes = {
	xs: 12,
	sm: 12,
	md: 12,
	lg: 11
};

export const localizedText = {
	locales: localized.locales,
	topBar: localized.topBar,
	login: localized.login,
	navigation: localized.navigation,
	pageNotFound: localized.pageNotFound
};
