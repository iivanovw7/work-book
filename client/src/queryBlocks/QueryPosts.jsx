import PropTypes from 'prop-types';
import React from 'react';
import { Query } from 'react-apollo';
import { withRouter } from 'react-router-dom';
import { ErrorMessage } from '../elements/errorMessage';
import { GET_POSTS } from '../graphql/queries';
import PostsList from '../components/Posts/PostsList';
import Spinner from '../elements/UI/Spinner';

const QueryPosts = props => (
  <Query query={GET_POSTS}>
    {({ loading, error, data }) => {
      if (loading) return <Spinner theme={props.theme} mgTop={5} />;
      if (error) {
        return (
          <ErrorMessage
            theme={props.theme}
            text="Back"
            message="Error =( Try again later..."
            handleClick={() => {
              props.history.push('/');
            }}
          />
        );
      }
      return <PostsList data={data} {...props} />;
    }}
  </Query>
);

export default withRouter(QueryPosts);

QueryPosts.propTypes = {
  history: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired
};
