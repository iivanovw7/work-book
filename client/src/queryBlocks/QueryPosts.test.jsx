import React from 'react';
import wait from 'waait';
import 'cross-fetch/polyfill';
import 'jsdom-global/register';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import * as mocks from '../__mocks__';
import * as testUtils from '../testUtils';
import QueryPosts from './QueryPosts';
// import 'babel-polyfill';

jest.mock('../config/apiURL', () => () => {
	return '/api';
});

jest.mock('../elements/UI/Button', () => () => (
	<div id="Button" />
));

describe('Testing QueryPosts() graphql query: ', () => {

	afterAll(() => {
		jest.clearAllMocks();
	});

	const queryProps = {
		locale: 'eng',
		text: testUtils.localizedText,
		theme: 'dark',
		history: testUtils.history
	};

	const Composition = (props) => {
		const { mocks } = props;
		return (
			<MemoryRouter>
				<MockedProvider mocks={mocks} addTypename={false} removeTypename>
					<QueryPosts {...queryProps} />
				</MockedProvider>
			</MemoryRouter>
		);
	};

	it('Should render Component and match snapshot', async () => {
		const component = mount(<Composition mocks={[mocks.gqlMocks[8]]} />);

		await wait(0);
		await component.update();
		expect(component).toMatchSnapshot();
	});

	it('Should render ErrorMessage', async () => {
		const component = mount(<Composition mocks={[mocks.gqlMocks[9]]} />);

		await wait(0);
		await component.update();
		expect(component.find('ErrorMessage').length).toEqual(1);
	});

	it('Should render Spinner', async () => {
		const component = mount(<Composition mocks={[]} />);

		await wait(0);
		expect(component.find('Spinner').length).toEqual(1);
	});

	it('Should render Component with correct data', async () => {
		const component = mount(<Composition mocks={[mocks.gqlMocks[8]]} />);

		await wait(0);
		await component.update();
		expect(component.find('PostsList').length).toEqual(1);
		expect(component.find('a').text()).toEqual('Custom Angular mouseWheel scroll directive');
	});
});
