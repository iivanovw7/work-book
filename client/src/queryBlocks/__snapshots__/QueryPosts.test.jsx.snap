// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing QueryPosts() graphql query:  Should render Component and match snapshot 1`] = `ReactWrapper {}`;
